### PR TITLE
Telecomms Refactor & CodeMirror

### PR DESCRIPTION
**Pull request by tigercat2000**
This commit does the following:
 - Quite a few new NTSL functions
  - Noteworthy changes:
  - implode() is technically handled now by vector2text; Actual implode() syntax kept for legacy.
  - A bunch of new math procs.
  - Lowercase true or false work now.
 - Absolute pathed telecomms scripting, all from https://github.com/d3athrow/vgstation13/pull/5713
 - Browser Datum traffic control
  - Absolutely lovely replacement for the fucking skin TCS window, using
    codemirror
 - CodeMirror JS, http://codemirror.net/
  - Custom compressed version, with CSS and PJB's NTSL syntax highlighting, search, sublime keymappings, and the base codemirror.
  - NanoUI integration, sorta, I didn't work on this as much as I wanted to, because IT TOOK
    11 FUCKING HOURS TO GET THE BROWSER DATUM TO WORK
  - Inspired by @PJB3005 's canceled [/vg/ todo list](https://github.com/d3athrow/vgstation13/pull/5713), probably not integrated even 1/8th as good as I'm sure that integration would have been.

Edit: One last thing, Something I have to point out.
>A while ago, #2233 added datastructures, including /datum/heap.
   The class /heap already existed in telecomms scripting code, LITERALLY THE ONLY FUCKING REASON IT COMPILED AT ALL IS BECAUSE BYOND TREATS `/datum/heap` AND `/heap` AS TWO FUCKING DIFFERENT TYPEPATHS, DESPITE THEM BEING EXACTLY THE DAMN SAME.

So, if any datastructure fuckups occur with this, Punch @Markolie  for me.
 

pretty
pretty
pretty

~~I chose CSS styling since it looked the prettiest out of all I tried, if anyone knows a more suitable codemirror supported styling, I'd be happy to integrate it.~~ This is now covered by PJB's custom NTSL syntax.

### CODE REVIEW

**Comment by Fox-McCloud** [comment]
10 bucks says this still doesn't prevent telecomms scripts from lagging the server to hell and back again =/

**Comment by tigercat2000** [comment]
Also, a section I *forgot* in the OP

### KNOWN BUGS:
 - Fucking cursor and text selection are out of line for everything following the first line, unfixable as far as I am concerned
 - The interface is a laggy piece of shit for some reason.

**Comment by Regen1** [comment]
>A lot of shit I am really too tired to fucking write about

Please follow the guidelines and include all changes inn this PR

**Comment by tigercat2000** [comment]
@Regen1 I'll update the OP later, before it's merged, this needs the 24 hour grace period anyways, it's primarily a feature.

**Comment by tigercat2000** [comment]
There, OP updated.

**Comment by PJB3005** [comment]
FYI, I made the actual codemirror syntax highlighter for NTSL, I just never got around to designing the UI fully & implementing it.

If you want it I've got it in a private (very messy) repo on GitLab.

**Comment by PJB3005** [comment]
>The interface is a laggy piece of shit for some reason.

I think this is because you need to turn on Edge mode, right now your UI is using IE fucking 6.

`<meta http-equiv="X-UA-Compatible" content="IE=edge"/>`

**Comment by tigercat2000** [comment]
@PJB3005 Ah, gotcha, I'll try that. A proper syntax highlighter would be great, too.

**Comment by PJB3005** [comment]
https://dl.dropboxusercontent.com/u/107028026/ShareX/2015/10/BLARG-master-e486193c0498db8dd1e01a01d64bf3e694312a10%20%281%29.zip

**Comment by tigercat2000** [comment]
@PJB3005 Cool, thanks.

File: nano/templates/code_mirror.tmpl [hunk 1]
```diff
@@ -0,0 +1,14 @@
+<h1>HELLO</h1>
```
Comment (PJB3005) [inline nano/templates/code_mirror.tmpl]: hrm

Comment (tigercat2000) [inline nano/templates/code_mirror.tmpl]: oh right that template, it was for testing the nanoUI elements
didn't remove it for anyone that every decided to fuck with nanoUI and codemirror

Comment (PJB3005) [inline nano/templates/code_mirror.tmpl]: NanoUI and codemirror won't work.

At all.

For the same reason you never got my chart to function in NanoUI.

HTML interface module for life.

Comment (tigercat2000) [inline nano/templates/code_mirror.tmpl]: I never developed it past making the codemirror window appear and take over the textbox- but it most assuredly works that far.

**Comment by tigercat2000** [comment]
New screenshot: screen

I noticed PJ had a custom theme for it as well, but it was giving me a difficult time properly implementing, and even then, I think the lesser dark looks nicer.

**Comment by tigercat2000** [comment]
yeah yeah

**Comment by Regen1** [comment]
Looks better

**Pull request merged**